hw/xbox/nv2a/nv2a_vsh.c

```diff
@@ -674,13 +674,24 @@ static const char* vsh_header =
     "#define EXP(dest, mask, src) dest.mask = _EXP(_in(src).x).mask\n"
     "vec4 _EXP(float src)\n"
     "{\n"
-    "  return vec4(exp2(src));\n"
+    "  vec4 result;\n"
+    "  result.x = exp2(floor(src));\n"
+    "  result.y = src - floor(src);\n"
+    "  result.z = exp2(src);\n"
+    "  result.w = 1.0;\n"
+    "  return result;\n"
     "}\n"
     "\n"
     "#define LOG(dest, mask, src) dest.mask = _LOG(_in(src).x).mask\n"
     "vec4 _LOG(float src)\n"
     "{\n"
-    "  return vec4(log2(src));\n"
+    "  float tmp = abs(src);\n"
```

Comment:
Should we specifically handle the case when `src == 0.0f` as the spec says?

Reply:
For the record: We don't handle the 0, INF, NAN cases anywhere right now because they are not as easy as it seems.

GLSL and NV VP floats differ in how those are treated, so by handling it, it's quite easy to accidentally introduce more errors. Look into NAN representation, FTZ and other things in the NV VP spec.

This is also one of the reasons I said I want to move VP into a CPU JIT in the future.

I'm fine with handling the 0.0 case here, but keep in mind that the proposed code is probably not a complete solution yet.